sklearn/datasets/tests/test_common.py

```diff
@@ -1,9 +1,40 @@
 """Test loaders for common functionality.
 """
+import pytest
+import numpy as np
+
+
+def check_pandas_dependency_message(fetch_func):
+    try:
+        import pandas  # noqa
+        pytest.skip("This test requires pandas to be not installed")
+    except ImportError:
+        # Check that pandas is imported lazily and that an informative error
+        # message is raised when pandas is missing:
+        expected_msg = ('{} with as_frame=True requires pandas'
+                        .format(fetch_func.__name__))
+        with pytest.raises(ImportError, match=expected_msg):
+            fetch_func(as_frame=True)
 
 
 def check_return_X_y(bunch, fetch_func_partial):
     X_y_tuple = fetch_func_partial(return_X_y=True)
     assert isinstance(X_y_tuple, tuple)
     assert X_y_tuple[0].shape == bunch.data.shape
     assert X_y_tuple[1].shape == bunch.target.shape
+
+
+def check_as_frame(bunch, fetch_func_partial):
+    pd = pytest.importorskip('pandas')
+    frame_bunch = fetch_func_partial(as_frame=True)
+    assert hasattr(frame_bunch, 'frame')
+    assert isinstance(frame_bunch.frame, pd.DataFrame)
+    assert isinstance(frame_bunch.data, pd.DataFrame)
+    assert frame_bunch.data.shape == bunch.data.shape
+    if frame_bunch.target.ndim > 1:
+        assert isinstance(frame_bunch.target, pd.DataFrame)
+    else:
+        assert isinstance(frame_bunch.target, pd.Series)
+    assert frame_bunch.target.shape[0] == bunch.target.shape[0]
+    assert frame_bunch.data.values.dtype == np.float64
```

Comment:
This check is not specific enough as pandas will automatically use float64 as the common dtype for the array view of the dataframe even if the original dtypes are distinct for different columns:

```python
>>> df = pd.DataFrame({"a": np.arange(3).astype(np.float32), 
...                    "b": np.arange(3).astype(np.float64), 
...                    "c": np.arange(3).astype(np.int32)})                                                                                                                     
>>> df.dtypes                                                                                                                                                                   
a    float32
b    float64
c      int32
dtype: object
>>> df.values.dtype                                                                                                                                                             
dtype('float64')
```

Here I would rather make the stricter check that the dtype for each column is always `np.float64`.